adapters/teads/teads.go

```diff
@@ -0,0 +1,178 @@
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	if len(request.Imp) == 0 {
+		return nil, []error{&errortypes.BadInput{
+			Message: "No impression in the bid request",
+		}}
+	}
+
+	endpointURL, err := a.buildEndpointURL()
+	if endpointURL == "" {
+		return nil, []error{err}
+	}
+
+	if err := updateImpObject(request.Imp); err != nil {
+		return nil, []error{err}
+	}
+
+	reqJSON, err := json.Marshal(request)
+	if err != nil {
+		return nil, []error{&errortypes.BadInput{
+			Message: "Error parsing BidRequest object",
+		}}
+	}
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	return []*adapters.RequestData{{
+		Method:  "POST",
+		Uri:     endpointURL,
+		Body:    reqJSON,
+		Headers: headers,
+	}}, []error{}
+}
+
+func updateImpObject(imps []openrtb2.Imp) error {
+	for i := range imps {
+		imp := &imps[i]
+
+		if imp.Banner != nil {
+			if len(imp.Banner.Format) != 0 {
+				bannerCopy := *imp.Banner
+				bannerCopy.H = &imp.Banner.Format[0].H
+				bannerCopy.W = &imp.Banner.Format[0].W
+				imp.Banner = &bannerCopy
+			}
+		}
+
+		var defaultImpExt DefaultBidderImpExtension
+		if err := json.Unmarshal(imp.Ext, &defaultImpExt); err != nil {
+			return &errortypes.BadInput{
+				Message: "Error parsing Imp.Ext object",
+			}
+		}
+		imp.TagID = strconv.Itoa(defaultImpExt.Bidder.PlacementId)
+		teadsImpExt := &TeadsImpExtension{
+			KV: TeadsKV{
+				PlacementId: defaultImpExt.Bidder.PlacementId,
+			},
+		}
+		if extJson, err := json.Marshal(teadsImpExt); err != nil {
+			return &errortypes.BadInput{
+				Message: "Error stringify Imp.Ext object",
+			}
+		} else {
+			imp.Ext = extJson
+		}
+	}
+	return nil
+}
+
+// Builds enpoint url based on adapter-specific pub settings from imp.ext
+func (a *adapter) buildEndpointURL() (string, error) {
+	endpointParams := macros.EndpointTemplateParams{}
+	host, err := macros.ResolveMacros(a.endpointTemplate, endpointParams)
+
+	if err != nil {
+		return "", &errortypes.BadInput{
+			Message: "Unable to parse endpoint url template: " + err.Error(),
+		}
+	}
+
+	thisURI, err := url.Parse(host)
+	if err != nil {
+		return "", &errortypes.BadInput{
+			Message: "Malformed URL: " + err.Error(),
+		}
+	}
+
+	return thisURI.String(), nil
+}
+
+func (a *adapter) MakeBids(internalRequest *openrtb2.BidRequest, _ *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+	if response.StatusCode == http.StatusBadRequest {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{fmt.Errorf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode)}
+	}
+
```

Comment:
just a suggestion - could make use of adapter utils here

https://github.com/prebid/prebid-server/blob/a292b3a356a0f7edeedf21a2d484a46bfb161064/adapters/response.go#L10-L28